### DISCUSSION TITLE
Some debug snippets frpm system crashes, database trouble, and so forth

### DISCUSSION THREAD

**Discussion by udippel** [gerbera/gerbera#2859] [General]
To whom it may concern:
Instead of filing bug reports, I'll add debug snippets from my experiences with gerbera. As a kind of repository for people who know the internal workings much better, and could have a better ides than myself about potential improvements.

```
2023-08-19 13:31:04 warning: Web Error: 300 Invalid username or password
2023-08-19 13:31:09 warning: Web Error: 300 Invalid username or password
2023-08-19 17:00:00 warning: Unrecognized action GetSortExtensionCapabilities
2023-08-19 17:00:00  error: ActionRequest::update(): response is nullptr, code -115 for GetSortExtensionCapabilities
2023-08-19 17:00:00 warning: Unrecognized action GetFeatureList
2023-08-19 17:00:00  error: ActionRequest::update(): response is nullptr, code -115 for GetFeatureList
2023-08-19 21:09:52   info: Gerbera shutting down. Please wait...
2023-08-19 21:09:52   info: Gerbera exiting. Have a nice day.
```
(Alas, with a crash. Fortunately no hanging.)

```
2023-08-20 15:42:00  error: Failed to execute script: Error: internal error
2023-08-20 15:42:00  error: [/build/src/content/scripting/script.cc:403] void Script::_execute() Error: Script: failed to execute script
```
(Repeatedly occurring at 'Add Item' )

**Comment by KarlStraussberger** [comment]
Thanks for that.

The two Get commands are interesting (but not mandatory) and not that complicated to implement. Only a bit of research in Upnp specs required.

Shutdown normally happens after SIGTERM is received - is there some message like that.

Also for the script error the context - file type, meta data settings - is of importance.

If you want to do further testing, try to get hold of the current head (might be available for ubuntu) because that is the reason it's not released yet.

**Comment by udippel** [comment]
For the Add Item, I have a serious hunch. It is not always happening, though whenever it happens, there was a file or m3u to be processed with that 'strange character', the Spanish "ñ". FYI: Using the web interface; so no interference by an app.

For the time being, I'll keep using and exploring the current ubuntu-deb

BubbleUPnP and Hi-Fi Cast seem to bring the database down without hesitation. However, Kodi and MX Player seem to generally work with gerbera. With Kodi being almost unusable, because it seems to be impossible to prevent it from shutting down playing when the power saving black screen sets in. Somehow it doesn't show in the Settings as a app that can be excluded from power saving (Samsung Android 13). So I stick to MX Player for the time being. If any other app is suggested, I will happily try that!

**Reply by KarlStraussberger** [reply]
The special character problem looks like a charset issue. What's your filesystem's charset value. That must be UTF-8 or set in config.xml

**Reply by KarlStraussberger** [reply]
Can you work out the exact version of your sqlite libraries. I tried BubbleUPnP and did not get a crash. Also it might be interesting to grab the first error/warning message before everything goes down.

**Comment by KarlStraussberger** [comment]
Just added a PR for the missing commands

**Comment by udippel** [comment]
To me it looked good. My keyboards are all swiss-german.
```
<scripting script-charset="UTF-8">
      <common-script>/usr/share/gerbera/js/common.js</common-script>
      <playlist-script>/usr/share/gerbera/js/playlists.js</playlist-script>
      <metafile-script>/usr/share/gerbera/js/metadata.js</metafile-script>
      <virtual-layout type="builtin">
        <import-script>/usr/share/gerbera/js/import.js</import-script>
      </virtual-layout>
    </scripting>

```

**Reply by KarlStraussberger** [reply]
The question is related to the settings on you file system. Filesystem charset can be configured separately.

**Reply by udippel** [reply]
"ISO-8859 text, with CRLF line terminators" in the ones that produce an error at 'Add Item'.

**Reply by KarlStraussberger** [reply]
Then you must set that in config.xml
```
<filesystem-charset>ISO-8859</filesystem-charset>
```
and maybe also
```
<metadata-charset>ISO-8859</metadata-charset>
```

(e.g. right befor `<scripting>`)

**Comment by udippel** [comment]
(Your question didn't show up here on github, yet I'll post the answer here:)

A deleted, freshly populated database. Started yesterday, and browsed a number of items with MX Player after 09:32:39; so the debug level doesn't seem to be very high.
 (I haven't deleted any lines in between)
Some minutes after 1 o'clock, started BubbleUPnP, Audio, Albums, select any, shows "15 tracks", press 'Play', dead. 

```
2023-08-21 09:32:39 warning: Web Error: 400 invalid session id
2023-08-21 13:09:02  error: prematurely shutting down.
2023-08-21 13:09:02  error: prematurely shutting down.
[/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-21 13:09:02  error: [/build/src/database/sqlite3/sqlite_database.cc:252] virtual std::shared_ptr<SQLResult> Sqlite3Database::select(const string&) Error: [/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-21 13:09:02  error: prematurely shutting down.
[/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-21 13:09:02  error: [/build/src/database/sqlite3/sqlite_database.cc:252] virtual std::shared_ptr<SQLResult> Sqlite3Database::select(const string&) Error: [/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-21 13:09:02  error: Exception: [/build/src/database/sqlite3/sqlite_database.cc:313] virtual int Sqlite3Database::exec(const string&, bool) Error: [/build/src/database/sqlite3/sl_task.cc:61] void SLTask::waitForTask() Error: SQLITE3 (19: 1555): UNIQUE constraint failed: grb_playstatus.group, grb_playstatus.item_id
       Query: INSERT INTO "grb_playstatus" ("group","item_id","playCount","lastPlayed","lastPlayedPosition","bookMarkPos") VALUES ('default',90396,1,1692616142,0,0)
       Error: UNIQUE constraint failed: grb_playstatus.group, grb_playstatus.item_id
2023-08-21 13:09:02  error: prematurely shutting down.
2023-08-21 13:09:02  error: [/build/src/database/sqlite3/sqlite_database.cc:277] virtual void Sqlite3Database::del(std::string_view, const string&, const std::vector<int>&) Error: [/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-21 13:09:02  error: prematurely shutting down.

```
.. and then it continues with the last two lines over and over. And it hangs, meaning the process keeps running but not responding, until I kill it. No further insight here:
```
2023-08-21 13:12:24 warning: Web Error: 800 Error: [/build/src/database/sqlite3/sqlite_database.cc:252] virtual std::shared_ptr<SQLResult> Sqlite3Database::select(const string&) Error: [/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-21 13:57:51   info: Gerbera shutting down. Please wait...
2023-08-21 13:57:51   info: Gerbera exiting. Have a nice day.

```

```
libsqlite3-0:amd64                    3.37.2-2ubuntu0.1                       amd64        SQLite 3 shared library
$ dpkg -L libsqlite3-0:amd64  
/.
/usr
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libsqlite3.so.0.8.6
/usr/share
/usr/share/doc
/usr/share/doc/libsqlite3-0
/usr/share/doc/libsqlite3-0/README.Debian
/usr/share/doc/libsqlite3-0/changelog.Debian.gz
/usr/share/doc/libsqlite3-0/copyright
/usr/lib/x86_64-linux-gnu/libsqlite3.so.0
```

**Reply by KarlStraussberger** [reply]
The sqlite is quite new (I'm on 3.39.3)

The hanging server results from all the database request blocking it from properly shutting down - it should not happen at all. I am already preparing a fix for the PlayStatus (update instead of delete and insert)

**Reply by udippel** [reply]
With this and the pull request as above integrated, I'd love a try out a new version in .deb.
"Production" here means a few users on a 2GB/8GB Thin Client. Compiling is out of question, and compiling elsewhere would not be production. 
It IS amazing, which tiny loads we get on this thin client. At least as long as there is no transcoding. Everything below 0.2, except of Add Item.

**Reply by KarlStraussberger** [reply]
There should be a gerbera_1.12.1+52~20230718201846-jammy1_amd64.deb on jfrog. It's not labeled `-edge` or so. The other ones will there as soon as the conan project has fixed a bug that blocks our CI pipeline.

**Reply by udippel** [reply]
Running ...

Will keep reporting any unintended behaviour.

**Reply by udippel** [reply]
As to be expected, alas, the database dies as it did before. I'd consider this a show stopper.

**Reply by KarlStraussberger** [reply]
Now we have gerbera_1.12.1+54~20230822234013-jammy1_amd64.deb with the fixes to the database code!

**Comment by udippel** [comment]
will do. What about the (only one) currently existing:
<scripting script-charset="UTF-8">?

**Reply by KarlStraussberger** [reply]
Normally you don't have to change that - It's the charset of the js files.

**Comment by udippel** [comment]
So I thought!

Alas, no cigar. The addition of 
`<filesystem-charset>ISO-8859</filesystem-charset>`
makes it a no-show: the file system is considered inaccessible  .. uh, whatever, through the web interface (clicking 'File System')
```
2023-08-21 17:23:57   info: The Web UI can be reached by following this link: http://192.168.1.220:49152
2023-08-21 17:24:20  error: iconv: Bad file descriptor
2023-08-21 17:24:20  error: iconv: Bad file descriptor
2023-08-21 17:24:20 warning: Web Error: 800 Error: [/build/src/util/string_converter.cc:170] std::string StringConverter::_convert(const string&, bool, std::size_t*) Error: iconv: Bad file descriptor
```
I have checked the diff between running and not running. I had added the two lines, first remove the metadata-one. Still failing. Then I removed the filesystem-line, and it started okay.

What next?

**Reply by KarlStraussberger** [reply]
My bad. The correct spelling is `ISO_8859-1`. Check out `iconv --list` for all available values.

**Reply by udippel** [reply]
No worries. My fault. I could have thought and checked myself!
Modified it, and now the file system is accessible. Have to wait for the reconstruction of the database ...

... no success. Same errors and skippings as before. 

I did read up in the meantime, and can confirm that my locale is UTF-8. As to be expected. I have, however, and I wasn't aware of this until just now, a bunch of m3u-files created on Windows (not by myself, beware!). 
convmv is no helper, because it doesn't convert files containing metadata. 

Here, on the GUI, loading these files look okay. On the gerbera box, opening them with cat, vi ... they do show the black rectangular question mark. 

So I guess, this problem isn't one to be attributed to gerbera. (Though I don't see any solution as of now.)

**Reply by udippel** [reply]
Thanks for the thumbs-up! 
At least, I'm learning something new. My active time was all and everything in English, much easier, and never really understood this UTF-thingy; except for my Chinese students and their need for UTF-16.

It IS unfortunate, that some of these char set mix-ups threw an error, other just 'skipping'. Also here, at the bottom of the wish-list, preferably some "bla-bla-bla has the wrong char set" in the log. Would, once again, save everyone's time.

By the way, this mixture is not to my taste. So in the end I found iconv and converted the bad guys to UTF-8 and again removed the 8859-1 from the config.xml. Now everything is in UTF-8 and no more errors and no more skipping.

**Reply by KarlStraussberger** [reply]
Actually, I used foobar to convert metadata charsets. And, with playlists you can do the same using dos2unix (rebase).

The problem is the iconv library, it just crashes from time to time. That's why we have a lot of options to set the charset - grep the config documentation. You may even use the tweaks dialog in folder view of the web ui...

**Comment by udippel** [comment]
While I'm gladly helping out, it might be the easiest to just make a 1:1 copy of my system. It doesn't run anything else, nothing special. If you happen to have an old box, or a VM: ubuntu server 22.04, all straightforward, except of language English, keyboard sg. Over. 
Add the jfrog-repository, install gerbera, set UI to yes, and we have parallel systems. 

As I wrote, I'd be more than happy to work with you; though it might just be simpler having this same system in front of you. Just as you like. 
Please, what could I do next?

**Reply by KarlStraussberger** [reply]
I'd love to do that. As a consequence I would have about 25 TV sets, 50 VMs and still no clue what the issue is, because I did things a bit different.
Normally, it's not that complicated to track down the issues and fix them.

**Reply by udippel** [reply]
All okay. So we keep working together. Looking forward to ...!

**Comment by udippel** [comment]
"and so forth", therefore:
`"warning: TagLibHandler /.../6_Channel_ID.wav: Does not handle the pcm content type"
`
Why? It is very minor, and yet, I don't understand why .wav-s et al should not be handled?

**Reply by KarlStraussberger** [reply]
That message comes from taglib and we only forward it. Obviously a file format issue here.

**Reply by udippel** [reply]
Obvious. I do understand that wav files may be without any tag. Though the message is actually stating something different: "we do not handle pcm type files". Actually, it should rather be something like "... _ID.wav: Does not contain any tag".
I'll contact the taglib people for this.

**Comment by udippel** [comment]
Awesome, thanks for the update!

**Reply by udippel** [reply]
One thumb up! One step further, good! Album plays. 
I purposely stop the playing; as if I was bored, done a search, alas dead again with:
```
2023-08-23 13:54:53 warning: Web Error: 400 invalid session id
2023-08-23 14:05:09  error: prematurely shutting down.
2023-08-23 14:05:09  error: Exception: [/build/src/database/sqlite3/sqlite_database.cc:313] virtual int Sqlite3Database::exec(const string&, bool) Error: [/build/src/database/sqlite3/sl_task.cc:61] void SLTask::waitForTask() Error: SQLITE3 (19: 1555): UNIQUE constraint failed: grb_playstatus.group, grb_playstatus.item_id
       Query: INSERT INTO "grb_playstatus" ("group","item_id","playCount","lastPlayed","lastPlayedPosition","bookMarkPos") VALUES ('default',71783,1,1692792309,0,0)
       Error: UNIQUE constraint failed: grb_playstatus.group, grb_playstatus.item_id
2023-08-23 14:05:09  error: prematurely shutting down.
[/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-23 14:05:09  error: [/build/src/database/sqlite3/sqlite_database.cc:252] virtual std::shared_ptr<SQLResult> Sqlite3Database::select(const string&) Error: [/build/src/database/sqlite3/sqlite_database.cc:407] void Sqlite3Database::addTask(const std::shared_ptr<SLTask>&, bool) Error: sqlite3 task queue is already closed
2023-08-23 14:05:09  error: prematurely shutting down.
[followed by some 10 repetitions of these last lines]
```
No resource problem, the tiny client has loads below 0.1.

Sorry for not much better news!

**Reply by KarlStraussberger** [reply]
Seems to be a very stubborn issue. This means that we have to accept the issue and avoid stopping gerbera because of it

**Reply by udippel** [reply]
I very much hope that you're joking. Neither with 'stopping gerbera' nor with 'accepting this behaviour'.
It might take some time to better shield database processes from one another, though. Since this is no real-time application, time is available for queuing, serialization, semaphores ... . Even dropping a request is a much lesser problem than a hung application. Any application, whenever it dies, ought to die gracefully. Cleanly. Could you agree?

**Reply by KarlStraussberger** [reply]
My idea is to ignore the error, because - as you mention - serializing the calls is quite complicated. Semaphores are out, because they are used on the call level already and we should not risk process blocks. Change is on its way.

**Reply by udippel** [reply]
To be honest: My test wasn't quite fair, since I had seen the code change, and noticed that the original code never had this type of clients (BubbleUPnP) in mind. Simple step-by-step clients like MX Player enforce _serialization by the user_: Select an item - play one item - stop this item - select another item - ... all fine. The new code would improve a lot and play an album, true.
Though - and here I was unfairly though correctly - expecting a similar concept for any higher level. I guessed to bring down the database with ease by doing a search within playing an album. You may call this multithreading. It isn't in its strict sense. I haven't seen the code, but I guess there is no 'command sanitizing level'. It wouldn't be the best of all worlds, but eventually the easiest way out, to abandon anything else issued earlier ('playing the album') and 'close' the database and instead execute the more recent search-command.  Real multithreading would be seriously fantastic, though also seriously difficult. Assuming a human being at the very end as a client, serial performance by 'one-after-the-other' would probably be acceptable. We humans are bad multi-taskers.

**Comment by udippel** [comment]
I have devised an idea to explain what I have in mind. I don't insist that it does what we need, but also cannot find any wrong in this concept. It will be necessary to draw apart potential multiple or even conflicting user input. 

Semaphore: Only one is needed. With a flag (F), like pedestrians' crossing. The system (Y) has basically two user inputs: Play (P) and Search (S). The semaphore has two states: Red (R) and Green (G). It controls the access of user input to Y. 
There is a queue store in front of the Semaphore, in case of multiple commands issued in short time (before Y can react).

If Y is doing whatsoever (P or S), Semaphore is set to R. Any user input sets F. Then Y closes any current action (P or S), Semaphore turns G and user input is accepted by Y. Semaphore turns R, and user input (P or S) is executed. Once S or P is finished, Semaphore turns G. 

If I am not mistaken, user input queue can be stack-ed. In case of multiple inputs, we can - if my assumption is correct and the user has changed his mind - make S perform (only) the last of the user's subsequent commands. 

With this intermediate layer of controlling user input, I would not have been able to bring the database down like I could easily some hours ago with "UNIQUE constraint failed", because it would have done only one singular thing. (Unfortunately playing the album would have stopped sooner or later. Though I guess that'd be acceptable.)

**Reply by KarlStraussberger** [reply]
First of all your idea will be a challenge to implement in that code, which in its core is about 20 years young. That's no excuse but technical debt to pay.

In fact, we have multi-threading at its best because of the http communication. Each call to gerbera is in its own thread. Otherwise we would not be able to serve multiple clients at the same time. That means, you're able to search and play by default. We always have to make sure that no device can block the server.

Furthermore, there is multi-threading on the database layer as well. The request is executed in another thread than the content manager code. Again, to avoid blocking the server with weird db searches.

Exactly, the multi-threading close to db level causes our issue: two select queries are ran with result: no entry (the original delete variant suffered from this even more). Then of course both threads want to insert: failure in the second. We tried to introduce transaction logic already to avoid this: sooner or later we ran into a deadlock. The code is still there but nobody activated this for a long time.

If asked for a way out: no way without completely rewriting the code. If asked when did the troubles start: when we introduced fast changing content like playstatus. My approach, instead of changing the world, I just ignore the error in cases I know it does no harm, like playstatus. That stops the server from terminating.

I wrote the staged import (config import-mode) this spring to be able to update faster. In the same way the database code can be refactored to bring the transaction logic into the first row. That will in the end apply the principles of blocking and non-blocking operations in transactions. A long way to go...

**Comment by udippel** [comment]
Thanks for the explanation.

Of course one needs multithreading for multiple sessions. Though I wasn't aware of multithreading within each session. That's of course better than I thought, and at the same time worse w.r.t what we encounter on the database level. Exactly as you wrote, when different threads can act differently if not contradictory on the same records, things behave like what can be observed.
I rest my (oversimplified) case.

**Reply by KarlStraussberger** [reply]
I just closed the PR and you can update to gerbera_1.12.1+56~20230823174413-jammy1_amd64.deb

**Reply by udippel** [reply]
Will do.
Thanks very much anyway! It has already shown a huge advantage that since yesterday we can at least nicely play albums, respectively a file followed by the subsequent ones.
